climada/entity/exposures/litpop/nightlight.py

```diff
@@ -222,53 +224,121 @@ def download_nl_files(req_files=np.ones(len(BM_FILENAMES),),
-            # this tile does not intersect the specified bounds
-            continue
-        extent = np.int64(np.clip(extent, 0, tile_size[None] - 1))
-
-        im_nl, _ = read_bm_file(SYSTEM_DIR, fname.replace('*', str(year)))
-        im_nl = np.flipud(im_nl)
-        im_nl = sparse.csc.csc_matrix(im_nl)
-        im_nl = im_nl[extent[0, 0]:extent[1, 0] + 1, extent[0, 1]:extent[1, 1] + 1]
-        nightlight.append((tile_coord, im_nl))
-
-    tile_coords = np.array([n[0] for n in nightlight])
-    shape = tile_coords.max(axis=0) - tile_coords.min(axis=0) + 1
-    nightlight = np.array([n[1] for n in nightlight]).reshape(shape, order='F')
-    nightlight = sparse.bmat(np.flipud(nightlight), format='csr')
-
-    coord_nl = np.vstack([coord_min, coord_h]).T
-    coord_nl[:, 0] += global_idx[0, :] * coord_h[:]
+    # open tif source file with raterio:
+    src = rasterio.open(path)
+    # read cropped data from  source file (src) to np.ndarray:
+    out_image, transform = rasterio.mask.mask(src, [geometry], crop=True)
+    meta = src.meta
+    meta.update({"driver": "GTiff",
+                 "height": out_image.shape[1],
+                 "width": out_image.shape[2],
+                 "transform": transform})
+    src.close()
+    return out_image[layer,:,:], meta
+
+# TODO: LitPop 2.0
+def load_nasa_nl_shape(geometry, reference_year, data_dir=None, dtype=None):
+    """Read nightlight data from NASA BlackMarble tiles
+    cropped to given shape(s) and combine.
+    
+    Parameters
+    ----------
+    geometry : shape(s) to crop data to in degree lon/lat.
+        for example shapely.geometry.(Multi)Polygon or shapefile.Shape.
+        from polygon defined in a shapefile. The object should have
+        attribute 'bounds' or 'points'
+    reference_year : int
+        target year for nightlight data, e.g. 2016.
+    data_dir : Path (optional)
+        Path to directory with BlackMarble data.
+        The default is SYSTEM_DIR.
+    dtype : dtype
+        data type for output default 'float32', required for LitPop,
+        choose 'int8' for integer.
+
+    Returns
+    -------
+    results_array : list containing numpy array
+    meta : list containing meta data per array
+    """
+    if data_dir is None:
+        data_dir = SYSTEM_DIR
+    if dtype is None:
+        dtype = 'float32'
+    if isinstance(geometry, Shape):
+        bounds = geometry.bbox
+    else:
+        bounds = geometry.bounds
+
+    # get closest available year from reference_year:
+    year = min(BM_YEARS, key=lambda x: abs(x - reference_year))
+
+    req_files = check_required_nl_files(bounds)
+    check_nl_local_file_exists(required_files=req_files, check_path=data_dir,
+                               year=year)
+    req_files = np.where(req_files ==1)[0] # convert to sorted list of indices
+    results_array_even = list() # tiles A1, B1, C1, D1 (Nothern Hemisphere)
+    results_array_odd = list() # tiles A2, B2, C2, D2 (Southern Hemisphere)
```

Comment:
?

Reply:
i added more comments here, this is all about combining the data from different black marble tiles into one array in the correct order (blackmarble data is provided in 8 tiles, each saved as TIFF.
see more new comments in code